stl/inc/memory

```diff
@@ -78,10 +78,22 @@ namespace ranges {
             _STL_INTERNAL_STATIC_ASSERT(_No_throw_sentinel_for<_OSe, _Out>);
             _STL_INTERNAL_STATIC_ASSERT(constructible_from<iter_value_t<_Out>, iter_reference_t<_It>>);
 
+            constexpr bool _Is_sized1 = sized_sentinel_for<_Se, _It>;
+            constexpr bool _Is_sized2 = sized_sentinel_for<_OSe, _Out>;
             if constexpr (_Ptr_copy_cat<_It, _Out>::_Really_trivial
-                          && sized_sentinel_for<_Se, _It> && sized_sentinel_for<_OSe, _Out>) {
-                return _Copy_memcpy_common(_IFirst, _RANGES next(_IFirst, _STD move(_ILast)), _OFirst,
-                    _RANGES next(_OFirst, _STD move(_OLast)));
+                          && (_Is_sized1 || same_as<_Se, unreachable_sentinel_t>) //
+#pragma warning(suppress : 6287) // Redundant code: the left and right subexpressions are identical
+                          &&(_Is_sized2 || same_as<_OSe, unreachable_sentinel_t>) ) {
+                if constexpr (_Is_sized1 && _Is_sized2) {
+                    return _Copy_memcpy_common(_IFirst, _RANGES next(_IFirst, _STD move(_ILast)), _OFirst,
+                        _RANGES next(_OFirst, _STD move(_OLast)));
+                } else if constexpr (_Is_sized1) {
+                    return _Copy_memcpy_distance(_IFirst, _OFirst, _IFirst, _RANGES next(_IFirst, _STD move(_ILast)));
+                } else if constexpr (_Is_sized2) {
+                    return _Copy_memcpy_distance(_IFirst, _OFirst, _OFirst, _RANGES next(_OFirst, _STD move(_OLast)));
+                } else {

```

Comment:
Ditto should this be a `_STATIC_ASSERT` and should we move this up to the top to detect this failure early?

Reply:
Ditto